docs/standard/microservices-architecture/multi-container-microservice-net-applications/implementing-api-gateways-with-ocelot.md

```diff
@@ -0,0 +1,579 @@
+
+All those services are currently implemented as ASP.NET Core Web API services, as you can tell because of the code. Let’s focus on one of the microservices like the Catalog microservice code.
+
+![](./media/image30.png)
+
+**Figure 8-29.** Sample Web API microservice (Catalog microservice)
+
+You can see that the Catalog microservice is a typical ASP.NET Core Web API project with several controllers and methods like in the following code.
+
+```csharp
+[HttpGet]
+[Route("items/{id:int}")]
+[ProducesResponseType((int)HttpStatusCode.NotFound)]
+[ProducesResponseType(typeof(CatalogItem),(int)HttpStatusCode.OK)]
+public async Task<IActionResult> GetItemById(int id)
+{
+    if (id <= 0)
+    {
+        return BadRequest();
+    }
+    var item = await _catalogContext.CatalogItems.
+                                          SingleOrDefaultAsync(ci => ci.Id == id);
+    //…
+
+    if (item != null)
+    {
+        return Ok(item);
+    }
+    return NotFound();
+}
+```
+The Http request will end up running that kind of C# code accessing the microservice database plus any additional action.
+
+In regards the microservice URL, when the containers are deployed in your local development PC (local Docker host), each microservice’s container has always an internal port, usually port 80, specified in its dockerfile, as in the following dockerfile.
+
+```csharp
+FROM microsoft/aspnetcore:2.0.5 AS base
+WORKDIR /app
+EXPOSE 80
+```
+The port 80 shown in the code is internal within the Docker host, so it cannot be reached by the client apps. 
+The client apps can access only to the external ports (if any) published when deploying with docker-compose.
+
+Those external ports should not be published when deploying into a production environment. This is precisely why you want to use the API Gateway, to avoid the direct communication between the client apps and the microservices.
+
+However, when developing, you want to access the microservice/container directly and run it through Swagger. That’s why in eShopOnContainers the external ports are still specified even when they won’t be used by the API Gateway or the client apps.
+
+Here’s an example of the docker-compose.override.yml file for the Catalog microservice.
+
+```csharp
+catalog.api:
+  environment:
+    - ASPNETCORE_ENVIRONMENT=Development
+    - ASPNETCORE_URLS=http://0.0.0.0:80
+    - ConnectionString=YOUR_VALUE
+    - ... Other Environment Variables
+  ports:
+    - "5101:80"   # Important: In a production environment you should remove the external port (5101) kept here for microservice debugging purposes. 
+                  # The API Gateway redirects and access through the internal port (80).
+```
+
+You can see how in the docker-compose.override.yml configuration the internal port for the Catalog container is port 80, but the port for external access is 5101. But this port shouldn’t be used by the application when using an API Gateway, only to debug run and test just the Catalog microservice.
```

Comment:
debug run and test -> debug, run, and test

Reply:
👍